lib/IDE/Refactoring.cpp

```diff
@@ -4757,6 +4755,163 @@ struct CallbackClassifier {
   }
 };
 
+class DeclCollector : private SourceEntityWalker {
+  llvm::DenseSet<const Decl *> &Decls;
+
+public:
+  /// Collect all explicit declarations declared in \p Scope (or \p SF if
+  /// \p Scope is empty) that are not within their own scope.
+  static void collect(BraceStmt *Scope, SourceFile &SF,
+                      llvm::DenseSet<const Decl *> &Decls) {
+    DeclCollector Collector(Decls);
+    if (!Scope) {
+      Collector.walk(SF);
+      return;
+    }
+
+    for (auto Node : Scope->getElements()) {
+      Collector.walk(Node);
+    }
+  }
+
+private:
+  DeclCollector(llvm::DenseSet<const Decl *> &Decls) : Decls(Decls) {}
+
+  bool walkToDeclPre(Decl *D, CharSourceRange Range) override {
+    if (!D->isImplicit())
+      Decls.insert(D);
+    return false;
+  }
+
+  bool walkToExprPre(Expr *E) override {
+    return !isa<ClosureExpr>(E);
+  }
+
+  bool walkToStmtPre(Stmt *S) override {
+    return !isa<BraceStmt>(S) && !isa<IfStmt>(S);
```

Comment:
Do we want to handle other kinds of statements that introduce a scope, e.g `do`, `while`, `switch`?